gh/client.py

```diff
@@ -91,10 +102,47 @@ def _events_since_repo_iter(self, date, repo):
         self.monitoring_db.write('events_count', { 'value': events }, tags={ 'repo': repo })
 
     def get_branches(self, repo):
-        return self.g.get_repo(repo).get_branches()
+        return self.get_repo(repo).get_branches()
 
     def get_pull_request(self, repo, num):
-        return self.g.get_repo(repo).get_pull(num)
+        return self.get_repo(repo).get_pull(num)
 
     def get_issue(self, repo, num):
-        return self.g.get_repo(repo).get_issue(num)
+        return self.get_repo(repo).get_issue(num)
+
+
+class ETagSupport:
+    def __init__(self, real_connection, storage):
+        self.real_connection = real_connection
+        self.storage = storage
+        self.logging = logging.getLogger('github_connection')
+
+    def can_use_etag(self, url):
+        return url.endswith('/pulls') or url.endswith('/events') or url.endswith('/issues')
+
+    def request(self, verb, url, input, headers):
+        previous_etag = None
+        if self.can_use_etag(url):
+            previous_etag = self.storage.load("ETag:%s" % url)
+            self.logging.info("---> %s %s (ETag: %s)" % (verb, url, previous_etag))
+            if previous_etag is not None:
+                headers['If-None-Match'] = previous_etag
+        else:
+            self.logging.info("---> %s %s (ETag: NotSupportedForUrl)" % (verb, url))
+
+        return self.real_connection.request(verb, url, input, headers)
+
+    def getresponse(self):
+        url = self.real_connection.url
+        response = self.real_connection.getresponse()
+        status = response.status
+        etag = response.headers['etag']
+        if self.can_use_etag(url) and etag is not None:
+            self.storage.store("ETag:%s" % url, etag)
+            self.logging.info("<--- %s (ETag: %s)" % (status, etag))
+        else:
+            self.logging.info("<--- %s" % status)
+        return response
```

Comment:
If this returns a 304, will the downstream logic be able to handle it?  (Ie, will it treat it just as an empty list, or is there any logic there that will fail on a non-2xx response?)

Reply:
yes, these are wrapped as PaginatedLists and it seems to treat it as an empty list.  I saw no problems in downstream logic.